libs/token/src/metadata.sw

```diff
@@ -0,0 +1,397 @@
+library;
+
+use src_7::Metadata;
+use std::{
+    bytes::Bytes,
+    hash::{
+        Hash,
+        sha256,
+    },
+    storage::{
+        storage_api::{
+            read,
+            write,
+        },
+        storage_bytes::*,
+        storage_string::*,
+    },
+    string::String,
+};
+
+/// A persistent storage type to store the SRC-7; Metadata Standard type.
+///
+/// # Additional Information
+///
+/// This type should only be used if metadata changes or cannot be set at contract deployment time.
+pub struct StorageMetadata {}
+
+impl StorageKey<StorageMetadata> {
+    /// Stores metadata for a specific asset and key pair.
+    ///
+    /// # Arugments
+    ///
+    /// * `asset`: [AssetId] - The asset for the metadata is to be stored.
+    /// * `key`: [String] - The key for the metadata is to be stored.
+    /// * `metadata`: [Metadata] - The metadata which is to be stored.
+    ///
+    /// # Number of Storage Accesses
+    ///
+    /// * Writes: `2`
+    ///
+    /// # Example
+    ///
+    /// ```sway
+    /// use src_7::Metadata;
+    /// use token::metdata::*;
+    ///
+    /// storage {
+    ///     metadata: StorageMetadata = StorageMetadata {}
+    /// }
+    ///
+    /// fn foo(asset: AssetId, key: String, metadata: Metadata) {
+    ///     storage.metadata.insert(asset, key, metadata);
+    /// }
+    /// ```
+    #[storage(read, write)]
+    pub fn insert(self, asset: AssetId, key: String, metadata: Metadata) {
+        let hashed_key = sha256((asset, key));
+
+        match metadata {
+            Metadata::Int(data) => {
+                write(hashed_key, 0, data);
+                write(sha256((hashed_key, self.slot)), 0, 0);
```

Comment:
Why does the last number in the second `write` increment for each metadata type while the first `write` seems to remain the same?

Doesn't the hashed key keep the collisions down to 0?

Reply:
The first write stores the data itself while the second write describes the type. This is needed as `String` and `Bytes` are pointer types and therefore cannot be stored, hence the use of the `StorageMetadata` type in the first place

Reply:
If you have a better suggestion as to how to handle this case I'm all ears